tensorflow/core/distributed_runtime/rpc/grpc_remote_worker.cc

```diff
@@ -260,10 +288,206 @@ class GrpcRemoteWorker : public WorkerInterface {
+    // std::cerr << response->DebugString() << std::endl;
+    RTresponse.ParseFromArray(&sta[4], sta.size() - sizeof(int));
+    // std::cerr << RTresponse.DebugString() << std::endl;
+
+    // Initialize the destination tensor
+    // TODO is it possible that this tensor already exists/reuse of memory
+    // location?
+    response->InitPartial(RTresponse);
+    const size_t nBytes = response->tensor().TotalBytes();
+
+    // Receive the Tensor content
+    char* data = const_cast<char*>(response->tensor().tensor_data().data());
+    for (size_t i = 0; i <= nBytes; i += maxSize) {
+      const int toRecv = std::min(maxSize, nBytes - i);
+      MPICheck(MPI_Recv(&data[i], toRecv, MPI_BYTE, src, followUpTag,
+                        MPI_COMM_WORLD, &status));
+    }
+#endif
+  }  // MPIRecvTensor
+
+  // Different MPI implementations use different values for allowed tag ranges
+  // so retrieve the max value to generate valid tags
+  int maxMessageTag;
+
+  // Max size of the data chuncks, 512MB
+  const size_t maxSize = 1024 * 1024 * 512;
+
+  // If the MPI_PATH_DISABLED variable is set then disable this code path
+  bool disabledAtRuntime = false;
+
+ public:
+  explicit RDMARemoteWorker(SharedGrpcChannelPtr channel,
+                            ::grpc::CompletionQueue* completion_queue,
+                            WorkerCacheLogger* logger)
+      : GrpcRemoteWorker(channel, completion_queue, logger) {
+    // Determine the maximum allowed message tag, used to determine hash
+    void* v;
+    int flag;
+    MPICheck(MPI_Comm_get_attr(MPI_COMM_WORLD, MPI_TAG_UB, &v, &flag));
+    maxMessageTag = *(int*)v;
+
+    // Test if this path is disabled at launch
+    const char* env = getenv("MPI_PATH_DISABLED");
+    if (env && env[0] == '1') disabledAtRuntime = true;
+  }
+
+  void SendTensorSync(const WorkerEnv* env, const Rendezvous::ParsedKey& key,
+                      const Rendezvous::Args& args, const Tensor& val,
+                      const bool is_dead, const int64 step_id_,
+                      Status& s) override {
+    const int dst = getMPIPartnerID(false, key.FullKey().ToString(), env);
+    const int hash =
+        tensorHash(strings::StrCat(key.FullKey().ToString(), step_id_));
+    // fprintf(stderr, "JBDBG Going to send data to: %s  : %d  || %d
+    // devContext: %d\n", key.FullKey().ToString().c_str(), dst, hash,
+    // args.device_context);
+    MPISendTensor(dst, hash, is_dead, val);
+    s = Status::OK();
+  }
+
+  void RecvTensorAsync(WorkerEnv* env, CallOptions* call_opts,
+                       const RecvTensorRequest* request,
+                       TensorResponse* response, StatusCallback done) override {
+    if (disabledAtRuntime) {
+      GrpcRemoteWorker::RecvTensorAsync(env, call_opts, request, response,
+                                        done);
+    } else {
+      // TODO: Figure out how to get the size of the requested Tensor
+      // Is this known by looking up the key/request?
```

Comment:
In a future version it will have the option to send the tensor in a single step, not for latency per se (although this can be configured if needed), but for handling certain tensor types that can not be trivially copied.

The fact that MPI takes care of all these things in MPI_Send/MPI_Recv is the main reason that I chose for the MPI API. TensorFlow is likely be used on a variety of machines and configurations and it would not be doable to implement it with verbs such that it will be optimal on all these networks. Better to let MPI implementations take care of that as they have been doing this for decades.